xarray/indexes/range_index.py

```diff
@@ -0,0 +1,366 @@
+        return {self.dim: positions}
+
+    def equals(self, other: CoordinateTransform) -> bool:
+        if not isinstance(other, RangeCoordinateTransform):
+            return False
+
+        return (
+            self.start == other.start
+            and self.stop == other.stop
+            and self.size == other.size
+        )
+
+    def slice(self, sl: slice) -> "RangeCoordinateTransform":
+        sl = normalize_slice(sl, self.size)
+
+        # TODO: support reverse transform (i.e., start > stop)?
+        assert sl.start < sl.stop
+
+        new_size = (sl.stop - sl.start) // sl.step
+        new_start = self.start + sl.start * self.step
+        new_stop = new_start + new_size * sl.step * self.step
+
+        return type(self)(
+            new_start, new_stop, new_size, self.coord_name, self.dim, dtype=self.dtype
+        )
+
+
+class RangeIndex(CoordinateTransformIndex):
+    """Xarray index implementing a simple bounded 1-dimension interval with
+    evenly spaced, monotonic floating-point values.
+
+    This index is memory-saving, i.e., the values of its associated coordinate
+    variable are not materialized in memory.
+
+    Do not use :py:meth:`~xarray.indexes.RangeIndex.__init__` directly. Instead
+    use :py:meth:`~xarray.indexes.RangeIndex.arange` or
+    :py:meth:`~xarray.indexes.RangeIndex.linspace`, which are similar to
+    :py:func:`numpy.arange` and :py:func:`numpy.linspace`.
+
+    In the case of a monotonic integer range, it is better using a
+    :py:class:`~xarray.indexes.PandasIndex` that wraps a
+    :py:class:`pandas.RangeIndex`.
+
+    """
+
+    transform: RangeCoordinateTransform
+
+    def __init__(self, transform: RangeCoordinateTransform):
+        super().__init__(transform)
+
+    @classmethod
+    def arange(
+        cls,
+        start: float = 0.0,
+        stop: float = 1.0,
+        step: float = 1.0,
+        *,
+        coord_name: Hashable | None = None,
+        dim: str,
+        dtype: Any = None,
+    ) -> "RangeIndex":
+        """Create a new RangeIndex from given start, stop and step values.
+
+        Parameters
+        ----------
+        start : float, optional
+            Start of interval (default: 0.0). The interval includes this value.
+        stop : float, optional
+            End of interval (default: 1.0). In general the interval does not
+            include this value, except floating point round-off affects the
+            size of the dimension.
+        step : float, optional
+            Spacing between values (default: 1.0).
+        coord_name : Hashable, optional
+            Name of the (lazy) coordinate variable that will be created and
+            associated with the new index. If ``None``, the coordinate is named
+            as the dimension name.
+        dim : str
+            Dimension name.
+        dtype : dtype, optional
+            The dtype of the coordinate variable (default: float64).
+
+        Examples
```

Comment:
```suggestion

        Note that all `start`, `stop` & `step` must be passed, which is more explicit than `np.arange` or `range`

        Examples
```

(optional, no strong view)

Reply:
> Note that all `start`, `stop` & `step` must be passed

This isn't exactly true, but yes the API here is more explicit than `np.arange` and `range`, e.g., `RangeIndex.arange(10.0)` means start=10 while `np.arange(10.0)` means stop=10.

`RangeIndex.arange(10.0)` doesn't make much sense, though, considering the default value of stop=1.0. I'll see if we can get closer to `np.arange` using `tpying.overload`.

Reply:
> `RangeIndex.arange(10.0)` doesn't make much sense, though, considering the default value of stop=1.0. I'll see if we can get closer to `np.arange` using `tpying.overload`.

yeah. no objection to the more explicit approach — it's useful-but-a-bit-magic that `arange` / `range` changes the meaning of the first arg based on how many are supplied

Reply:
Mimicing `numpy.arange` behavior is surprisingly difficult! (at least for me, I've been struggling with this).

I got it close with some simple logic, but then I hit the same issue than https://github.com/numpy/numpy/pull/17878 (i.e., `RangeIndex.arange(start=10)` returns a range in the [0, 10) interval, which makes no sense). I could fix it after some heavy refactor making the code / API ugly to a point I'm not sure I want to push it here :).

Numpy relies on the Python C API `PyArg_ParseTupleAndKeywords()` but AFAIK there seem to be no easy way to know from Python whether a value has been passed as positional or keyword argument.

Reply:
!!

I think the explicit approach is very valid, but maybe we just call it out / ensure people need to pass kwargs where it could be confusing

Reply:
In [4a128d0](https://github.com/pydata/xarray/pull/10076/commits/4a128d0fdaee6639336ff900ecdf4b6e3a613fa2) I've chosen to mimicking numpy and pandas API anyway, with some simple logic and clearly documenting the caveat above.

I find myself (likely others too) doing `range(10)` or `np.arange(10.0)` so many times while I doubt many will write something like `np.arange(start=10)`.

`pandas.RangeIndex(start=10)` actually still returns `RangeIndex(start=0, stop=10, step=1)` and I haven't seen anyone complaining in the pandas issues (or I missed it).